lib/engines/csvEngine.js

```diff
@@ -0,0 +1,163 @@
+    this.simulator = simulator;
+    this.mqttServer = "tcp://" + simulator.configuration.server;
+    this.dataGenerators = {};
+}
+
+CSVEngine.prototype.generateValues = function(dataDescriptor, system) {
+    var self = this;
+    var result = {};
+
+    var filename = this.config.fileRepository + "/" + system.name + ".csv";
+    console.log("    Reading data [" + filename + "]");
+    var config = this.config;
+    return new Bromise(function(resolve, reject) {
+        fs.readFile(filename, 'utf8', function(err, data) {
+            if (err) {
+                reject(err);
+            }
+            converter.csv2json(data, function(err, json) {
+                if (err) {
+                    console.log(err);
+                    reject(err);
+                }
+
+                try {
+                    // Browse each data
+                    console.log("    Generating", json.length, "data");
+                    json.forEach(function(object) {
+                        var timestamp = object[self.getTimestampConfig(dataDescriptor)] + "000";
+                        result[timestamp] = {};
+                        result[timestamp][dataDescriptor] = object[self.getDataValueConfig(dataDescriptor)];
+                    });
+                    resolve(result);
+                } catch (e) {
+                    console.error("# Data gen error:", e);
+                    reject(e);
+                }
+            }, config.csvOptions);
+        });
+    });
+};
+
+CSVEngine.prototype.generateData = function(system) {
+    var self = this;
+    console.log("Generate data for system:", system.name);
+    //var datas = {};
+    try {
+        return _.reduce(self.getGenerationData(), function(datas, dataDescriptor, dataName) {
+            console.log("generate value for data:", dataName);
+
+            return self.generateValues(dataName, system).then(function(data) {
+                return _.extend(datas, data);
+            });
+        }, {});
+
+    } catch (e) {
+        console.log(e);
+    }
+
+
+    //return datas;
+    /*var generateAllDatas = new Bromise.reduce(self.getGenerationData(), function(datas, dataDescriptor, dataName) {
+        console.log("generate value for data:", dataName);
+        return self.generateValues(dataName, system).then(function(data) {
+            return _.extend(datas, data);
+        });
+    }, {});
+
+    generateAllDatas.then(function(datas) {
+        console.log(datas);
+    });*/
+};
+
+CSVEngine.prototype.getGenerationData = function() {
+    return this.simulator.simulation.generation.data;
+};
+
+CSVEngine.prototype.getDataValueConfig = function(dataId) {
+    return this.simulator.simulation.generation.data[dataId].value;
+};
+
+CSVEngine.prototype.getTimestampConfig = function(dataId) {
+    return this.simulator.simulation.generation.data[dataId].timestamp;
+};
+
+CSVEngine.prototype.sendData = function(system) {
+
+    var self = this;
```

Comment:
Again, arrow functions allows us to get rid of these `var self = this;` and do:

```js
try {
   return this..generateData(system).then(data => {
        return new Bromise((resolve, reject) => {
              //...
              const mqttClient = createMQTTClient(system, this.mqttServer);
              //...
         });
    });
}
```

Reply:
Removed var self= this,
Replaced:
        return self.generateData(system).then(function(data) {
            return new Bromise(function(resolve, reject) {
by your code
change self by this in mqttClient call